botocove/cove_decorator.py

```diff
@@ -80,3 +83,17 @@ def wrapper(*args: Any, **kwargs: Any) -> CoveOutput:
         return decorator
     else:
         return decorator(_func)
+
+
+def _validate_ids(ids: Optional[List[str]]) -> None:
```

Comment:
Minor org nitpick - maybe this makes more sense within the CoveHostAccount init than in the decorator mechanism

More useful thought -  We could make these str into concrete classes earlier, so that in _list_accounts_for_resource there isn't a fail case on line 169 that can raise ValueError and we can is_instance() at that point instead of _is_regex() checks. The regex could live in that class too perhaps so we can fail fast before listing the org accounts.

It'd also make our type hints cleaner: resource_ids would be List[Union[AccountIdStr, OrgOuStr]] , account_map as Dict[AccountIdStr, AccountTypeDef]

Reply:
My thinking here was to check the types as early as possible, since Python lets the user pass any old object despite what the type hint says.

I'm fine with moving it back to `CoveHostAccount` but not sure where to put it. Would you put it in `_list_accounts_for_resource` or somewhere else?

If there hadn't already been a test for this particular validation (`test_malformed_ignore_ids_type`) I would probably have ignored this part and just let duck typing happen at runtime.

To get serious about validation and domain types, would you consider using a library like [Pydantic](https://pydantic-docs.helpmanual.io/) to handle it? It seems like a big dependency to include at runtime but it's designed to handle this sort of thing.

(An aside: the validation I would really like to see is checking that `regions` is a list of strings and not a single string! So many times I have gotten errors when trying to access the regions `e`, `u`, `-`, `w`, and so on :-) )

Reply:
I'm pretty deep into the Pydantic/FastAPI world at $job and it definitely inspired that line of thinking to have validators, but Pydantic is way too big a dependency to pull in to a library, so it's something I'm set on ruling out in this case. Having being burned by libs like `cryptography` before, I'd rather our tree is as simple as it can be.

I think we can easily get the same value out of just using stricter data types internally (but leaner than the iteration of Botocove that went a bit overboard on Dataclasses - I enjoyed learning about the dunder copy and deepcopy stuff from that adventure!) - but happy to patch that later rather than scope creep you here. 

Feel free to forget this and I'll pick up another time :)

> (An aside: the validation I would really like to see is checking that regions is a list of strings and not a single string! So many times I have gotten errors when trying to access the regions e, u, -, w, and so on :-) )

absolutely - I get bitten by this (and similarly set("str") == {"s", "t", "r"} reasonably often enough to be a bit ashamed of it :) The worst is when mypy would catch it but hasn't been run yet to point out the error and a confused debugging session ends with an "I told you so" from typecheck :D